kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatcherWebSocketListener.java

```diff
@@ -48,11 +48,14 @@ public void onError(WebSocket webSocket, Throwable t) {
 
   @Override
   public void onMessage(WebSocket webSocket, String text) {
-    webSocket.request();
     // onMesssage and onClose are serialized, but it's not specified if onError
     // may occur simultaneous with onMessage.  So we prevent concurrent processing
-    synchronized (this) {
-      manager.onMessage(text);
+    try {
+      synchronized (this) {
+        manager.onMessage(text);
+      }
+    } finally {
+      webSocket.request();
```

Comment:
Why the try-finally? Won't this force a new request even if the user interrupts the execution, or there's any other sort of runtime exception?

Reply:
I see the new test verifies this, but I'm not really sure why we wouldn't want the message processing to stop in **any** event

Reply:
> Why the try-finally? Won't this force a new request even if the user interrupts the execution, or there's any other sort of runtime exception?
> but I'm not really sure why we wouldn't want the message processing to stop in any event

That's never been part of the contract.  Any Exception (including runtime) thrown by the Watcher is caught and the Watch continues.  You could make the case for an unchecked exception causing Watch termination, but that will be a minor breaking change.

This test is just ensuring that if that catch is inadvertently removed, we'll still continue with message processing.

Reply:
For me the test is fine(, and also the behavior). I was just pointing this out because I'm unaware of the implementation parts of webSocket.request, and how might that be affected.

We should also provide (if not there), similar tests to those we implemented for the Portforward to verify that the message processing is performed in order + that a new requests are sent only when the previous message has been consumed.

Reply:
> I'm unaware of the implementation parts of webSocket.request, and how might that be affected.

A call to request is required to receive the next message or onClose (this matches the documented behavior of the jdk client).  If you call it after onClose or onError - it will be a no-op as those will be terminal events.

Reply:
> If you call it after onClose or onError - it will be a no-op as those will be terminal events.

Should we add this (or a similar description) as part of the interface's Javadoc?

Reply:
> Should we add this (or a similar description) as part of the interface's Javadoc?

Sure I'll add something.